{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPr5RUMhKoLCLkUA3q0ukwx"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# BF-931 TÓPICOS AVANÇADOS EM MICOLOGIA I - Introdução ao Biopython\n","\n","## Programa de Pós-Graduação em Biologia de Fungos - [PPGBF](https://www.ufpe.br/ppgbf)\n","\n","### [**Dra. Angelina de Meiras-Ottoni**](http://lattes.cnpq.br/5692217174749691)\n","\n","**LinkedIn:** [Angelina Meiras Ottoni](https://www.linkedin.com/in/angelina-meiras-ottoni/)\n","\n","**GitHub:** [AngelOttoni](https://github.com/AngelOttoni)\n","\n","[Source repository](https://github.com/AngelOttoni/learning-biopython)"],"metadata":{"id":"IDRs6CHHm_Jb"}},{"cell_type":"markdown","source":["# **Python libraries**\n","\n","- In Python, libraries  are collections of pre-written code and functionalities that extend the language's capabilities beyond its core features.\n","\n","- These libraries are developed to address specific tasks and domains, such as data manipulation, scientific computing, machine learning, web development, and more.\n","\n","- They are created to save developers time and effort by providing ready-to-use functions and tools, making it easier to implement complex tasks without having to write everything from scratch.\n","\n","- Python's extensive library ecosystem is one of its major strengths, as it allows developers to leverage existing solutions to common problems rather than reinventing the wheel.\n","\n","- By importing and using these libraries in their code, developers can access a wide range of functionalities, making their programs more powerful, efficient, and feature-rich.\n","\n","Libraries in Python can be distributed in different ways:\n","\n","1. **[Standard Library:](https://docs.python.org/3/library/)** Python comes with a comprehensive standard library, which includes modules for handling various tasks, such as file I/O, regular expressions, networking, data structures, and more.\n","\n","  - These modules are available by default and do not require additional installation.\n","\n","2. **Third-Party Libraries:** Apart from the standard library, there are numerous third-party libraries developed by the Python community and other organizations.\n","  - These libraries cover diverse areas like data analysis, web development, artificial intelligence, game development, and more.\n","  - To use these libraries, developers need to install them using package managers like [pip](https://pip.pypa.io/en/stable/) [(Python Package Installer)](https://packaging.python.org/en/latest/tutorials/installing-packages/).\n","\n","**Here's an example of importing and using a library in Python:**"],"metadata":{"id":"ITd5yDHCnM-X"}},{"cell_type":"code","source":["# Importing the NumPy library\n","import numpy as np\n","\n","# Creating a NumPy array\n","my_array = np.array([1, 2, 3, 4, 5])\n","\n","# Performing a calculation with NumPy\n","result = np.sum(my_array)\n","\n","# Printing the result\n","print(result)"],"metadata":{"id":"l5HDriUmp8OS"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["- In this example, we imported the [NumPy library](https://numpy.org/) using the `import` statement and used its functions and capabilities to create an array and calculate its sum efficiently.\n","\n","- Overall, libraries in Python play a vital role in expanding the language's functionality, making it a versatile and powerful choice for a wide range of applications.\n","\n","- Developers can explore and leverage different libraries based on their project requirements to streamline development and enhance their code's capabilities."],"metadata":{"id":"mXpZaXVFqGwq"}},{"cell_type":"markdown","source":["# **Biopython**\n","\n","\n","\n","---\n","\n","- [Biopython](https://biopython.org/) is a powerful and popular open-source Python library designed to facilitate bioinformatics and computational biology tasks.\n","\n","- It provides a wide range of functionalities for handling biological data, including DNA, RNA, protein sequences, and various biological file formats.\n","\n","- Biopython is a valuable resource for researchers and bioinformaticians to analyze, manipulate, and process biological data efficiently."],"metadata":{"id":"0FYBkj46qgMZ"}},{"cell_type":"markdown","source":["## **History**\n","\n","\n","\n","---\n","\n","\n","\n","1. **Early Development (1999-2000):**\n","   Biopython originated in 1999 when **Jeff Chang**, a graduate student at the University of California, San Diego, started the project as a collection of Python modules to work with biological data. Initially, it was a personal project for Jeff's research needs.\n","\n","2. **Project Expansion (2000-2006):**\n","   As the project gained popularity, Jeff decided to **release Biopython under an open-source license in 2000**, making it freely available for the bioinformatics community.\n","   \n","    - Over the next few years, the community started to grow, and more contributors joined the project.\n","    - Biopython expanded its scope to encompass a broader range of bioinformatics tasks.\n","\n","3. **Biopython Tutorial (2003):**\n","   **Brad Chapman**, one of the contributors, created the **first Biopython tutorial in 2003**, which made it easier for users to learn and get started with the library.\n","\n","4. **Stable Releases (2004-2010):**\n","   The Biopython project started to achieve stability with regular releases, making it more reliable for research and analysis.\n","   \n","    - The library incorporated functionalities like sequence handling, file parsing, sequence analysis, and phylogenetics.\n","\n","5. **Integration of BioSQL (2005-2006):**\n","   Biopython **integrated with BioSQL**, a database schema designed for storing and querying biological data.\n","\n","6. **Continued Growth (2010-2015):**\n","   Biopython continued to grow, attracting more contributors from various backgrounds, including academia, industry, and the **open-source** community.\n","\n","7. **Python 3 Support (2015-2016):**\n","   With the increasing importance of Python 3, Biopython underwent significant changes to ensure compatibility and support for Python 3, along with maintaining support for Python 2.\n","\n","8. **Modernization and PEP 8 Compliance (2017-2018):**\n","   Biopython focused on code modernization and improved adherence to the PEP 8 style guide, making the codebase cleaner and easier to maintain.\n","\n","9. **Beyond the 1.70 Release (2019-2021):**\n","   The Biopython project continued to progress, with efforts on performance optimization, bug fixes, and new feature additions. The community remained active, supporting users and encouraging contributions."],"metadata":{"id":"7K58nE_b2DW2"}},{"cell_type":"markdown","source":["**Here's a brief introduction to some of the key features and functionalities of Biopython:**\n","\n","\n","---\n","\n","\n","\n","1. **Sequence Handling:**\n","   Biopython allows you to work with biological sequences, such as DNA, RNA, and protein sequences.\n","    - It provides data structures to store and manipulate these sequences, as well as methods for accessing individual elements, calculating sequence statistics, and performing sequence alignments.\n","\n","2. **File Parsing:**\n","   The library supports reading and writing various biological file formats, such as FASTA, GenBank, PDB (Protein Data Bank), Clustal, and many more.\n","    - This capability is essential for loading data from external sources and saving results in a standardized format.\n","\n","3. **Sequence Analysis:**\n","   Biopython offers a wide range of tools for sequence analysis, including functions for sequence similarity searches, motif finding, ORF (Open Reading Frame) prediction, translation, reverse complement, and more.\n","   \n","    - It also supports various algorithms for sequence alignment, such as BLAST and ClustalW.\n","\n","4. **Phylogenetics:**\n","   Biopython enables you to perform phylogenetic analyses, such as building phylogenetic trees, calculating evolutionary distances, and conducting bootstrap analysis.\n","   \n","    - It supports popular tools like Bio.Phylo for phylogenetic workflows.\n","\n","5. **Structural Biology:**\n","   The library provides functionalities to work with 3D protein structures, including parsing PDB files, analyzing protein structure properties, and performing structural alignments.\n","\n","6. **Bioinformatics Utilities:**\n","   Biopython includes several utilities for handling biological data, such as tools for handling codon tables, calculating molecular weights, and handling genetic codes.\n","\n","**To start using Biopython, you first need to [install](https://biopython.org/wiki/Download) it using `pip`:**"],"metadata":{"id":"QAHfB1rRnJlb"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"pnKWKmSWm05U"},"outputs":[],"source":["!pip install biopython"]},{"cell_type":"markdown","source":["- Once installed, you can import the Biopython modules in your Python script and begin exploring the library's capabilities."],"metadata":{"id":"ddXqj3OQsKjQ"}},{"cell_type":"code","source":["# Remenber: upload file"],"metadata":{"id":"DpErlNqnxbHV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from Bio import SeqIO\n","\n","# Example: Reading a FASTA file\n","filename = \"falename.fasta\"\n","for record in SeqIO.parse(filename, \"fasta\"):\n","    print(\"ID:\", record.id)\n","    print(\"Sequence:\", record.seq)"],"metadata":{"id":"6FLg6mxlsIKx"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## **Some necessary clarifications:**\n","\n","\n","---\n","\n","\n","\n","1. **Module:** A module is indeed a set of functions, types, classes, or variables grouped together in a common namespace.\n","\n","  - It is a single file containing Python code and can be imported and used in other Python programs.\n","\n","2. **Library:** A library is a collection of modules (and sometimes additional resources) that are designed to be used in various programs.\n","\n","  - It typically contains reusable code that provides specific functionalities. Libraries are not necessarily confined to Python; they can be written in other programming languages as well.\n","\n","3. **Package:** A package is a way of organizing related modules and sub-packages in a directory hierarchy.\n","\n","  - It is essentially a directory that contains a special file called `__init__.py`, which indicates that the directory is a package.\n","  \n","  - A package can include multiple modules and sub-packages, making it a higher-level organizational concept in Python.\n","\n","**So, to summarize:**\n","\n","\n","- A **module** is a single file containing Python code.\n","\n","- A **library** is a collection of modules (and potentially additional resources) that offer specific functionality and can be used in various programs.\n","\n","- A **package** is a directory that contains one or more modules and sub-packages, indicated by the presence of the `__init__.py` file, used for organizing related functionality.\n","\n","- **Regarding the term** \"*Package*: Distribution unit that can contain a library, an executable, or both,\" this definition might be referring to a specific type of package in the context of packaging and distribution, which is not the same as the Python package discussed above.\n","\n","- In the context of packaging, a package could refer to a distribution unit that contains the necessary files to distribute and install a library or an executable.\n","\n","- This distribution unit may include metadata, configuration files, documentation, and code, allowing users to easily install and use the software on their systems."],"metadata":{"id":"YrJyLN0I5K0i"}},{"cell_type":"code","source":["print(\"The end\")"],"metadata":{"id":"TMEkLDjz60CP"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Reference links:\n","\n","\n","---\n","\n","[Python](https://www.python.org/)\n","\n","[Bio package](https://biopython.org/docs/1.75/api/Bio.html)\n","\n","[Biopython](https://biopython.org/)\n","\n","[Open Bioinformatics Foundation](https://www.open-bio.org/)\n","\n","[Biostars](https://www.biostars.org/)\n","\n","[Stack overflow Bioinfo](https://stackoverflow.com/questions/tagged/biopython)"],"metadata":{"id":"GPOFBei_1Y5H"}}]}